plugins/parquet-store/parquet_store.cpp

```diff
@@ -0,0 +1,722 @@
+    for (auto& slice : batch) {
+      if (self->state.vast_type_) {
+        VAST_ASSERT(*self->state.vast_type_ == slice.layout());
+      } else {
+        self->state.vast_type_ = slice.layout();
+      }
+      self->state.num_rows_ += slice.rows();
+      self->state.table_slices_.push_back(slice);
+    }
+    VAST_TRACE("{} received batch of {} table slices for partition {}", *self,
+               batch.size(), self->state.id_);
+  };
+}
+
+std::shared_ptr<parquet::WriterProperties>
+writer_properties(const configuration& config) {
+  auto builder = parquet::WriterProperties::Builder{};
+  builder.created_by("VAST telemetry engine")
+    ->max_row_group_length(detail::narrow_cast<int64_t>(config.row_group_size))
+    ->enable_dictionary()
+    ->compression(parquet::Compression::ZSTD)
+    ->compression_level(9)
+    ->version(parquet::ParquetVersion::PARQUET_2_6);
+  return builder.build();
+}
+
+// TODO: make this static or something - no need to recompute every time
+std::shared_ptr<parquet::ArrowWriterProperties> arrow_writer_properties() {
+  auto builder = parquet::ArrowWriterProperties::Builder{};
+  builder.store_schema(); // serialize arrow schema into parquet meta data
+  return builder.build();
+}
+
+auto make_import_time_col(const time& import_time, int64_t rows) {
+  auto v = import_time.time_since_epoch().count();
+  auto builder = time_type::make_arrow_builder(arrow::default_memory_pool());
+  if (auto status = builder->Reserve(rows); !status.ok())
+    die(fmt::format("make time column failed: '{}'", status.ToString()));
+  for (int i = 0; i < rows; ++i) {
+    auto status = builder->Append(v);
+    VAST_ASSERT(status.ok());
+  }
+  return builder->Finish().ValueOrDie();
+}
+
+auto create_record_batch(const table_slice& slice)
+  -> std::shared_ptr<arrow::RecordBatch> {
+  auto rb = to_record_batch(slice);
+  auto md = rb->schema()->metadata();
+  auto event_array = rb->ToStructArray().ValueOrDie();
+  auto time_col = make_import_time_col(slice.import_time(), rb->num_rows());
+  auto schema = arrow::schema(
+    {arrow::field("import_time", time_type::to_arrow_type()),
+     arrow::field("event", event_array->type(), rb->schema()->metadata())});
+  auto new_rb
+    = arrow::RecordBatch::Make(schema, rb->num_rows(), {time_col, event_array});
+  return new_rb;
+}
+
+auto write_parquet_buffer(const std::vector<table_slice>& slices,
+                          configuration config) {
+  auto sink = arrow::io::BufferOutputStream::Create().ValueOrDie();
+  auto batches = arrow::RecordBatchVector{};
+  for (const auto& slice : slices)
+    batches.push_back(create_record_batch(slice));
+  auto table = arrow::Table::FromRecordBatches(batches).ValueOrDie();
+  auto writer_props = writer_properties(config);
+  auto arrow_writer_props = arrow_writer_properties();
+  auto status
+    = parquet::arrow::WriteTable(*table, arrow::default_memory_pool(), sink,
+                                 1 << 24, writer_props, arrow_writer_props);
```

Comment:
Is the 2^24 here the row group size?

Reply:
Can you add a comment that describes this? Otherwise someone else will have no clue how the magic numbers have been chosen.